### PR TITLE
Walkthroughs

### PR DESCRIPTION
**Pull request by angelinechen9**
I extended my servlet to support full pagination and deletion of specific comments.  I implemented user authentication.

Because I think that using JavaScript to manipulate the DOM is not the best way to display entries from the database, I changed some of the code in the examples.  I improved the structure of the code by changing HTML to JSP.

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -41,18 +45,36 @@
...
-        response.setContentType("application/json;");
-        response.getWriter().println(gson.toJson(comments));
+        request.setAttribute("comments", comments);
+        UserService userService = UserServiceFactory.getUserService();
+        if (!userService.isUserLoggedIn()) {
+            String loginUrl = userService.createLoginURL("/data");
+            request.setAttribute("loginStatus", false);
+            request.setAttribute("loginUrl", loginUrl);
+        }
+        else {
```
Comment (Caffeinix) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: The Google Java style guide for if/else says to put the else on the same line as the preceding close brace.  I don't know if that's the style guide you're trying to follow here though. :)

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -41,18 +45,36 @@
...
     @Override
-    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
         Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
         DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
         PreparedQuery results = datastore.prepare(query);
         List<Comment> comments = new ArrayList<>();
         for (Entity entity : results.asIterable()) {
-            Comment comment = new Comment(entity.getKey().getId(), (String) entity.getProperty("comment"), (long) entity.getProperty("timestamp"));
+            Comment comment = new Comment(entity.getKey().getId(), (String) entity.getProperty("nickname"), (String) entity.getProperty("comment"), (long) entity.getProperty("timestamp"));
```
Comment (Caffeinix) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: You might want to wrap this long line.  Most auto-formatters for Java (including the ones in most IDEs) will let you specify a wrap column.

File: portfolio/src/main/webapp/nickname.jsp [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.2.0/css/all.css" integrity="sha384-hWVjflwFxL6sNzntih27bfxkr27PmbbK/iSvJ+a4+0owXq79v+lsFkW54bOGbiDQ" crossorigin="anonymous">
+    <script src="https://code.jquery.com/jquery-3.3.1.slim.min.js" integrity="sha384-q8i/X+965DzO0rT7abK41JStQIAqVgRVzpbzo5smXKp4YfRvH+8abtTE1Pi6jizo" crossorigin="anonymous"></script>
+    <script src="script.js"></script>
+  </head>
+  <body>
+    <nav class="navbar navbar-expand-lg navbar-light bg-white">
+        <a class="nav-link" href="/data">Back</a>
+    </nav>
+    <main>
+        <form action="/nickname" method="POST">
```
Comment (Caffeinix) [inline portfolio/src/main/webapp/nickname.jsp:20]: This seems to be mostly a duplicate of the form in data.jsp.  What's the purpose of each of these?  Without a screenshot or a way to see your site, it's hard for me to understand what's going on here.  Maybe some HTML comments would help?

Comment (angelinechen9) [reply portfolio/src/main/webapp/nickname.jsp:20]: The form in data.jsp creates a comment.  The form in nickname.jsp changes the nickname.  I commented the code in the servlet that renders the templates.  Should I also comment the code in the templates?

Comment (Caffeinix) [reply portfolio/src/main/webapp/nickname.jsp:20]: That would be really helpful, yes.  It looks like you can use `<%-- --%>` to add JSP comments that will be stripped when the page is rendered.

**Review by Caffeinix — APPROVED** [review]

**Pull request merged**